discovery_engine_core/core/src/stack/filters/semantic.rs

```diff
@@ -16,40 +16,39 @@ use std::collections::BTreeMap;
 
 use itertools::{izip, Itertools};
 use kodama::{linkage, Dendrogram, Method};
-use xayn_discovery_engine_ai::{nan_safe_f32_cmp, pairwise_cosine_similarity};
+use xayn_discovery_engine_ai::{l2_norm, nan_safe_f32_cmp, triangular_product};
 
 use crate::document::{Document, WeightedSource};
 
 use super::source_weight;
 
 /// Computes the condensed cosine similarity matrix of the documents' embeddings.
 fn condensed_cosine_similarity(documents: &[Document]) -> Vec<f32> {
-    pairwise_cosine_similarity(
-        documents
-            .iter()
-            .map(|document| document.smbert_embedding.view()),
-    )
-    .indexed_iter()
-    .filter_map(|((i, j), &similarity)| (i < j).then(|| similarity))
+    triangular_product(documents, |doc_a: &Document, doc_b: &Document, _i, _j| {
```

Comment:
Tried to refactor this, but the test fails, the original code is a bit complicated, would love to better understand it